go-controller/pkg/factory/handler.go

```diff
@@ -500,56 +533,33 @@ func newQueuedInformer(oType reflect.Type, sharedInformer cache.SharedIndexInfor
 	if err != nil {
 		return nil, err
 	}
-	i.events = make([]chan *event, numEventQueues)
-	i.shutdownWg.Add(len(i.events))
-	for j := range i.events {
-		i.events[j] = make(chan *event, 10)
-		go i.processEvents(i.events[j], stopChan)
-	}
+	i.queueMap = newQueueMap(numEventQueues, &i.shutdownWg)
+	i.queueMap.start(stopChan)
+
 	i.initialAddFunc = func(h *Handler, items []interface{}) {
 		// Make a handler-specific channel array across which the
 		// initial add events will be distributed. When a new handler
 		// is added, only that handler should receive events for all
 		// existing objects.
-		type initialAddEntry struct {
-			obj      interface{}
-			doneFunc func()
-		}
-		adds := make([]chan *initialAddEntry, numEventQueues)
-		queueWg := &sync.WaitGroup{}
-		queueWg.Add(len(adds))
-		for j := range adds {
-			adds[j] = make(chan *initialAddEntry, 10)
-			go func(addChan chan *initialAddEntry) {
-				defer queueWg.Done()
-				for {
-					entry, ok := <-addChan
-					if !ok {
-						return
-					}
-					h.OnAdd(entry.obj)
-					entry.doneFunc()
-				}
-			}(adds[j])
-		}
+		addsWg := &sync.WaitGroup{}
+		addsMap := newQueueMap(numEventQueues, addsWg)
+		addsMap.start(stopChan)
+
 		// Distribute the existing items into the handler-specific
 		// channel array.
 		for _, obj := range items {
-			key, entry := i.refQueueEntry(i.oType, obj, numEventQueues)
-			adds[entry.queue] <- &initialAddEntry{
-				obj: obj,
-				doneFunc: func() {
-					i.unrefQueueEntry(key, entry, false)
-				},
-			}
-		}
-		// Close all the channels
-		for j := range adds {
-			close(adds[j])
+			key, entry := addsMap.refEntry(i.oType, obj)
+			addsMap.enqueueEvent(nil, obj, entry.queue, func(e *event) {
+				h.OnAdd(e.obj)
+				addsMap.unrefEntry(key, entry, false)
+			})
 		}
+
 		// Wait until all the object additions have been processed
-		queueWg.Wait()
+		addsMap.shutdown()
```

Comment:
I find current shutdown/stopChan design a bit confusing. there are 2 ways to stop events processing
1. close the channel, and no more events will be handled
2. call shutdown and all events that are already in the queues will be processed before exiting

so maybe we should rename `shutdown` to something like `handleReceivedEventsAndExit` or any name that will explain what is going on

also I think it makes sense to either pass `stopChan` to `newInformer` and ` newQueuedInformer` and remove `(i *informer) shutdown()`, or the opposite, create internal `stopChannel` for informers and only have  `(i *informer) shutdown()`, otherwise you can close eventQueues with stop channel, but still have handlers running

Reply:
@npinaeva I agree, it's confusing. We actually want to do both 1 & 2 in all cases (both test-cases and regular execution) but unfortunately we cannot do (1) until we upgrade our client-go code to 1.26 which finally has https://github.com/kubernetes/kubernetes/pull/112200 that will allow us to correctly shut down our informers in our testcases. For regular execution it doesn't matter much because we're exiting anyway.

But even better, when we have https://github.com/kubernetes/kubernetes/pull/112200 I believe we can probably remove a bunch of the factory code entirely and just use filtered shared informers instead (where factory is a thin wrapper) because the only reason for the factory to exist was that up until 112200 you couldn't ever shut down a shared informer, which we need to do when eg NetworkPolicies get deleted.

So maybe all this code goes away soonish?

Reply:
sounds good, the only thing I didn't get is how we want 1 & 2, if one of them says handle events that are in the queue and exit, and another one says don't handle any more events and exit? for me (1) sounds like a real shutdown, and (2) like a special case for the initial add events handling?